### PR TITLE
Enable/disable listeners

### PR DESCRIPTION
**Pull request by ddey2**
Addressed:
1. Adding a page for listing all the extractors
2. Only admins should be able to enable/disable them.
3. Regular users should see the enabled extractors only.

To test:
1. run the test_extractors
2. Try to switch to admin_mode and then enable/disable the extractors
3. Then try to switch back to regular mode and you should see only enabled extractors.
4. Try to disable an extractor, then try to submit a job being a regular user from backend(Postman), you should see Forbidden error.

Adding some screenshots for reference.

Screenshot 2024-05-02 at 5 16 16 PM
Screenshot 2024-05-06 at 3 38 38 PM

### CODE REVIEW

**Review by tcnichol — APPROVED** [review]
Tested and everything works. Marking approved.

**Comment by Vismayak** [comment]
Tested and working, approved!

**Review by Vismayak — APPROVED** [review]

File: backend/app/routers/listeners.py [hunk 1]
```diff
@@ -374,10 +391,66 @@ async def edit_listener(
...
+    return await set_active_flag(listener_id, True)
+
+
+@router.put("/{listener_id}/disable", response_model=EventListenerOut)
+async def disable_listener(
+    listener_id: str,
+    user_id=Depends(get_user),
+    allow: bool = Depends(ListenerAuthorization()),
+):
+    """Enable an Event Listener. Only admins can enable listeners.
```
Comment (lmarini) [inline backend/app/routers/listeners.py:414]: disable

File: backend/app/routers/listeners.py [hunk 2]
```diff
@@ -374,10 +391,66 @@ async def edit_listener(
...
+
+    Arguments:
+        listener_id -- UUID of the listener to be enabled
```
Comment (lmarini) [inline backend/app/routers/listeners.py:417]: diabled

File: backend/app/routers/listeners.py [hunk 3]
```diff
@@ -374,10 +391,66 @@ async def edit_listener(
...
+    """
+    return await set_active_flag(listener_id, False)
+
+
+@router.put("/{listener_id}/toggle", response_model=EventListenerOut)
```
Comment (lmarini) [inline backend/app/routers/listeners.py:422]: if this toggles, do we need the other two methods above?

**Review by lmarini — CHANGES_REQUESTED** [review]
Besides some minor comments there is a couple of high level issues I noticed:

1. Anyone can click on the extractor link, but if you are not in admin mode, nothing shows in the list. Maybe we can have a message for now. Or not show the link in the menu unless you are in admin. I am not sure.
2. Once I went into admin, I clicked on two extractors, but when I go to the dataset submission page, three show active. I countinued to click and the lists on dataset page and file page of enabled extractors are off from the main page.

Thanks!

File: backend/app/models/listeners.py [hunk 4]
```diff
@@ -69,6 +69,7 @@ class EventListenerDB(Document, EventListenerBase):
     modified: datetime = Field(default_factory=datetime.now)
     lastAlive: datetime = None
     alive: Optional[bool] = None  # made up field to indicate if extractor is alive
+    active: Optional[bool] = None
```
Comment (lmarini) [inline backend/app/models/listeners.py:72]: This should probably be just `bool` and not optional. It's either enabled or not. Default should be False.

(Same could be said about `alive` above)

**Review by lmarini — CHANGES_REQUESTED** [review]

File: backend/app/deps/authorization_deps.py [hunk 5]
```diff
@@ -389,6 +390,43 @@ async def __call__(
...
+                listener.creator and listener.creator.email == current_user
+            ):
+                return True
+            else:
+                raise HTTPException(
+                    status_code=403,
+                    detail=f"User `{current_user} does not have permission on listener `{listener_id}`",
+                )
+        raise HTTPException(
+            status_code=404, detail=f":Listener {listener_id} not found"
```
Comment (lmarini) [inline backend/app/deps/authorization_deps.py:426]: Can we remote the starting `:`?

**Comment by ddey2** [comment]
> Besides some minor comments there is a couple of high level issues I noticed:
> 
> 1. Anyone can click on the extractor link, but if you are not in admin mode, nothing shows in the list. Maybe we can have a message for now. Or not show the link in the menu unless you are in admin. I am not sure.
> 2. Once I went into admin, I clicked on two extractors, but when I go to the dataset submission page, three show active. I countinued to click and the lists on dataset page and file page of enabled extractors are off from the main page.
> 
> Thanks!

Hi Luigi,

1. I agree. I was thinking of hiding the link in the menu unless you are an admin and not in admin_mode.
 2. I am not able to reproduce that. I am trying with 2 extractors. It reflects only the enabled extractors on my file/dataset page.

File: backend/app/deps/authorization_deps.py [hunk 6]
```diff
@@ -389,6 +390,43 @@ async def __call__(
         raise HTTPException(status_code=404, detail=f"Group {group_id} not found")
 
 
+class ListenerAuthorization:
+    """We use class dependency so that we can provide the `permission` parameter to the dependency.
+    For more info see https://fastapi.tiangolo.com/advanced/advanced-dependencies/.
+    We want not admins users not to access listeners which are not active"""
```
[the remark this answers is not in the document]
Comment (ddey2) [reply backend/app/deps/authorization_deps.py:396]: done

File: backend/app/routers/listeners.py (hunk 3 above)
Comment (ddey2) [reply backend/app/routers/listeners.py:422]: I am using this from frontend to toggle the flag. I was thinking if there is a use-case where somebody explicitly calls a backend endpoint to enable/disable the listener irrespective of the active flag. Hence, I created the separate endpoints, but all use the same function internally.

Comment (longshuicy) [reply backend/app/routers/listeners.py:422]: image

After I select the extractors, what do I do to enable/disable? I don't see any button around. I should already be in admin mode.

File: backend/app/routers/listeners.py [hunk 7]
```diff
@@ -374,10 +391,66 @@ async def edit_listener(
...
+@router.put("/{listener_id}/disable", response_model=EventListenerOut)
+async def disable_listener(
+    listener_id: str,
+    user_id=Depends(get_user),
+    allow: bool = Depends(ListenerAuthorization()),
+):
+    """Disable an Event Listener. Only admins can enable listeners.
+
+    Arguments:
+        listener_id -- UUID of the listener to be enabled
```
Comment (longshuicy) [inline backend/app/routers/listeners.py:417]: These two endpoints seems a little bit redundant? Could it be just one endpoint accepting the parameter then decide to activate or de-activate

File: backend/app/routers/listeners.py [hunk 8]
```diff
@@ -374,10 +391,66 @@ async def edit_listener(
...
+):
+    """Toggle the active flag of an Event Listener. Only admins can enable/disbale listeners.
+
+    Arguments:
+        listener_id -- UUID of the listener to be enabled
+    """
+    listener = await EventListenerDB.find_one(
+        EventListenerDB.id == ObjectId(listener_id)
+    )
+    if listener:
```
Comment (longshuicy) [inline backend/app/routers/listeners.py:436]: in the dependency class `ListenerAuthorization`, You might be already checking the id exist?

File: backend/app/routers/listeners.py (hunk 3 above)
Comment (ddey2) [reply backend/app/routers/listeners.py:422]: @longshuicy Clicking on the checkboxes does the job. In this case, both the extractors are enabled.

File: backend/app/routers/listeners.py (hunk 7 above)
Comment (ddey2) [reply backend/app/routers/listeners.py:417]: Please see my reply to Luigi above. 

If you all think this is redundant, I will get rid of these two.

File: backend/app/routers/listeners.py (hunk 8 above)
Comment (ddey2) [reply backend/app/routers/listeners.py:436]: I don't if admin and admin_mode both are true

**Comment by ddey2** [comment]
May 16th meeting -
1.  try toggle button instead of checkbox 
2. List all the extractors, disable toggle button for nonadmin mode
3. make set_active_flag method as private and let other endpoints call them

**Comment by ddey2** [comment]
Major update:
1. Use toggle switch button instead of checkbox
2. Showing tooltip for disabled toggle button

For latest UI look, please refer to ss

Screenshot 2024-05-16 at 2 05 16 PM
Screenshot 2024-05-16 at 12 23 04 PM

**Review by lmarini — APPROVED** [review]
Looks great! Thanks for making those changes.

**Pull request merged**